lib/compiler/src/beam_ssa_opt.erl

```diff
@@ -1479,41 +1489,89 @@ do_opt_try([{L,Blk}|Bs]=Bs0, Ws0) ->
     [].
 
-is_safe_without_try([#b_set{op=kill_try_tag}|Is], Acc) ->
+reduce_try_is([#b_set{op=kill_try_tag}|Is], Acc) ->
     %% Remove this kill_try_tag instruction. If there was a landingpad
     %% instruction in this block, it has already been removed. Preserve
     %% all other instructions in the block.
     {done,reverse(Acc, Is)};
-is_safe_without_try([#b_set{op=extract}|_], _Acc) ->
+reduce_try_is([#b_set{op=extract}|_], _Acc) ->
     %% The error reason is accessed.
     unsafe;
-is_safe_without_try([#b_set{op=landingpad}|Is], Acc) ->
-    is_safe_without_try(Is, Acc);
-is_safe_without_try([#b_set{op={succeeded,body}}=I0|Is], Acc) ->
+reduce_try_is([#b_set{op=landingpad}|Is], Acc) ->
+    reduce_try_is(Is, Acc);
+reduce_try_is([#b_set{op={succeeded,body}}=I0|Is], Acc) ->
     %% If we reached this point, it means that the previous instruction
     %% has no side effects. We must now convert the flavor of the
     %% succeeded to the `guard`, since the try/catch will be removed.
     I = I0#b_set{op={succeeded,guard}},
-    is_safe_without_try(Is, [I|Acc]);
-is_safe_without_try([#b_set{op=Op}=I|Is], Acc) ->
+    reduce_try_is(Is, [I|Acc]);
+reduce_try_is([#b_set{op=Op}=I|Is], Acc) ->
     IsSafe = case Op of
                  phi -> true;
                  _ -> beam_ssa:no_side_effect(I)
              end,
     case IsSafe of
-        true -> is_safe_without_try(Is, [I|Acc]);
+        true -> reduce_try_is(Is, [I|Acc]);
         false -> unsafe
     end;
-is_safe_without_try([], Acc) ->
+reduce_try_is([], Acc) ->
     {safe,reverse(Acc)}.
 
+%% Removes try/catch expressions whose expressions will never throw.
+trim_try([{L, #b_blk{is=[#b_set{op=new_try_tag,dst=Tag}],
+                     last=Last0}=Blk0} | Bs],
+         UnreachablePads0, KilledTags0, Acc) ->
+    #b_br{succ=SuccLbl,fail=PadLbl} = Last0,
+    case cerl_sets:is_element(PadLbl, UnreachablePads0) of
+        true ->
+            %% The landing pad can't be reached in any way, remove the entire
+            %% try/catch.
+            Blk = Blk0#b_blk{is=[],
+                             last=#b_br{bool=#b_literal{val=true},
+                                        succ=SuccLbl,fail=SuccLbl}},
+
+            UnreachablePads = cerl_sets:del_element(PadLbl, UnreachablePads0),
+            KilledTags = cerl_sets:add_element(Tag, KilledTags0),
+            trim_try(Bs, UnreachablePads, KilledTags, [{L, Blk} | Acc]);
+        false ->
+            trim_try(Bs, UnreachablePads0, KilledTags0, [{L, Blk0} | Acc])
+    end;
+trim_try([{L, #b_blk{is=[#b_set{op=landingpad} | _]}=Blk}| Bs],
+         UnreachablePads0, KilledTags, Acc) ->
+    UnreachablePads = cerl_sets:add_element(L, UnreachablePads0),
+    trim_try(Bs, UnreachablePads, KilledTags, [{L, Blk} | Acc]);
+trim_try([{L, Blk} | Bs], UnreachablePads, KilledTags, Acc) ->
+    Successors = beam_ssa:successors(Blk),
+    Pads = maps:without(Successors, UnreachablePads),
```

Comment:
What does Dialyzer think about this? (You are treating a cerl_set, which is opaque, as a map.)

Reply:
Fixed locally.